src/System.Text.Json/tests/Serialization/Stream.WriteTests.cs

```diff
@@ -258,6 +263,632 @@ public static async Task LargeJsonFile(int bufferSize)
             }
         }
+
+        [Theory]
+        [InlineData(1, true, true)]
+        [InlineData(1, true, false)]
+        [InlineData(1, false, true)]
+        [InlineData(1, false, false)]
+        [InlineData(10, true, false)]
+        [InlineData(10, false, false)]
+        [InlineData(100, true, false)]
+        [InlineData(1000, true, false)]
+        public static async Task VeryLargeJsonFileTest(int payloadSize, bool ignoreNull, bool writeIndented)
+        {
+            List<Post> list = PopulateLargeObject(payloadSize);
+
+            JsonSerializerOptions options = new JsonSerializerOptions
+            {
+                IgnoreNullValues = ignoreNull,
+                WriteIndented = writeIndented
+            };
+
+            string json = JsonSerializer.Serialize(list, options);
+
+            // Sync case.
+            {
+                List<Post> deserializedList = JsonSerializer.Deserialize<List<Post>>(json, options);
+                Assert.Equal(payloadSize, deserializedList.Count);
+
+                string jsonSerialized = JsonSerializer.Serialize(deserializedList, options);
+                Assert.Equal(json, jsonSerialized);
+            }
+
+            // Async case.
+            using (var memoryStream = new MemoryStream())
+            {
+                await JsonSerializer.SerializeAsync(memoryStream, list, options);
+                string jsonSerialized = Encoding.UTF8.GetString(memoryStream.ToArray());
+                Assert.Equal(json, jsonSerialized);
+
+                memoryStream.Position = 0;
+                List<Post> deserializedList = await JsonSerializer.DeserializeAsync<List<Post>>(memoryStream, options);
+                Assert.Equal(payloadSize, deserializedList.Count);
+            }
+        }
+
+        [Theory]
+        [InlineData(1, true, true)]
+        [InlineData(1, true, false)]
+        [InlineData(1, false, true)]
+        [InlineData(1, false, false)]
+        [InlineData(2, true, false)]
+        [InlineData(2, false, false)]
+        [InlineData(4, true, false)]
+        [InlineData(8, true, false)]
+        [InlineData(16, true, false)]
+        public static async Task DeepNestedJsonFileTest(int depthFactor, bool ignoreNull, bool writeIndented)
+        {
+            int length = 10 * depthFactor;
+            List<Post>[] posts = new List<Post>[length];
+            posts[0] = PopulateLargeObject(1);
+            for (int i = 1; i < length; i++ )
+            {
+                posts[i] = PopulateLargeObject(1);
+                posts[i - 1][0].Value.PrimaryTopic.RelatedTopics = posts[i];
+            }
+
+            JsonSerializerOptions options = new JsonSerializerOptions()
+            {
+                MaxDepth = depthFactor * 64,
+                IgnoreNullValues = ignoreNull,
+                WriteIndented = writeIndented
+            };
+            string json = JsonSerializer.Serialize(posts[0], options);
```

Comment:
It is test for deep nested object. Array itself not used for serialization because first element within array is parent of all other array elements. i.e. 1st element is parent of 2nd, 2nd is parent of 3rd, 3rd is parent of 4th etc. Used array to keep reference to objects and assign next element as child of previous.  So even only first post from array used for serialization the array elements are used entirely in nested objects.